### PR TITLE
[2/n] Consolidate `replicate` and `DDP`: split `forward` function

### PR DESCRIPTION
**Pull request by yhcharles**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #96658

Split `forward` function into `pre_forward` and `post_forward`, so that they can be reused in the composable API of `replicate`.

Differential Revision: [D44377456](https://our.internmc.facebook.com/intern/diff/D44377456)

### CODE REVIEW

File: torch/nn/parallel/distributed.py [hunk 1]
```diff
@@ -1542,6 +1535,17 @@ def forward(self, *inputs, **kwargs):
         self._clear_grad_buffer()
         return output
 
+    def forward(self, *inputs, **kwargs):
+        with torch.autograd.profiler.record_function("DistributedDataParallel.forward"):
+            if self._delay_all_reduce_all_params:
```
Comment (rohan-varma) [inline torch/nn/parallel/distributed.py:1540]: why don't we move this logic to pre-forward?

File: torch/nn/parallel/distributed.py [hunk 2]
```diff
@@ -1542,6 +1535,17 @@ def forward(self, *inputs, **kwargs):
...
+                output = self.module.forward(*inputs, **kwargs)
+                self._clear_grad_buffer()
```
Comment (rohan-varma) [inline torch/nn/parallel/distributed.py:1542]: why is this needed only for delay AR?

File: torch/nn/parallel/distributed.py [hunk 3]
```diff
@@ -2115,10 +2119,7 @@ def _check_comm_hook(self, hook):
                 "Communication hook: return annotation should be torch.futures.Future[torch.Tensor].",
             )
 
-        if hook.__name__ in [
-            "bf16_compress_hook",
-            "bf16_compress_wrapper_hook",
-        ] and (
+        if hook.__name__ in ["bf16_compress_hook", "bf16_compress_wrapper_hook"] and (
```
Comment (rohan-varma) [inline torch/nn/parallel/distributed.py:2122]: nit: unrelated change?

**Review by rohan-varma — COMMENTED** [review]
just a small q, thanks!

File: torch/nn/parallel/distributed.py (hunk 2 above)
Comment (yhcharles) [reply torch/nn/parallel/distributed.py:1542]: Thanks for pointing out. I think I can move `self._clear_grad_buffer()` to `_post_forward()` as well. Let me update.
PS: The motivation is to move all operations before `self.module.forward()` into `_pre_forward()`, and operations after `self.module.forward()` into `_post_forward()`. So that `_pre_forward()` and `_post_forward()` can be reused in composable APIs.

**Review by rohan-varma — COMMENTED** [review]
LGTM overall, since this is a large change, can we run our internal QPS benchmarks and confirm no regression? Thanks!

**Comment by yhcharles** [comment]
@yhcharles has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D44377456).

**Review by rohan-varma — APPROVED** [review]
looks good, thanks for running benchmark and helping the consolidation!

**Comment by yhcharles** [comment]
Benchmark done. No regresison for QPS or accuracy. Results can be found in https://www.internalfb.com/diff/D44377456

**Comment by yhcharles** [comment]
@yhcharles has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D44464781).

**Comment by yhcharles** [comment]
@yhcharles has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D44377456).
